sky/serve/load_balancer.py

```diff
@@ -18,25 +21,135 @@
+#
+# - Automated Unit/Smoke Testing: Obviously this needs to be implemented before
+# merging.  For now the code was tested by hand.
+#
+# - Hitless Configuration Changes: When configured with a yaml file, Envoy must
+# be restarted to take up configuration changes.  When done naively (as the
+# PR currently does), restarting will cause all existing connections to be
+# dropped, forcing clients to retry later.  As written, this means that any
+# time a replica is added or removed, connections to all replicas will be
+# impacted.  There are two ways to deal with this:
+#   - Hot Restart:  Envoy has a mechanism for "hot restart".   When used, a new
+#   Envoy boots and begins handling connections with the new coniguration while
+#   the old one drains.
+#   - XDS:  Envoy has a robust remote update protocol that allows changes to be
+#   made to it's configuration at runtime.  XDS support may be needed anyways
+#   if we move to having multiple Envoy Load balancers that are not necessarily
+#   co-located with the controller.  As a result it may be worth skipping hot
+#   restart and building this out instead.
+#
+# - Autoscaling:  PythonLoadBalancer uses a list of each request and the
+# time it was handled to calculate a qps number which the controller uses to
+# make autoscaling decisions.  Envoy doesn't export stats with this level of
+# detail, so we will need to find a different approach.  A couple of
+# considerations:
+#    - We could simulate the existing functionality by simply asking Envoy how
+#    many requests it has handled for each replica on some regular cadence, and
+#    using that to calculate an estimated qps number.   This should be
+#    relatively straight forward to implement.
+#    - We should consider, however, whether qps is really the right way to be
+#    making auto-scaling decisions.  For AI serving workloads, does qps really
+#    tell you how loaded the replica is?   In the non-AI world, one would
+#    generally use CPU/Memory pressure for this purpose, however there are some
+#    challenges here with regard to measuring load on GPUs.  This requires
+#    further discussion.
+#
+# - Single Envoy:  In this PR we boot an Envoy per service, mostly to be
+# consistent with the approach the PythonLoadBalancer takes.   This approach
+# has two major problems (1) given that we could have several Envoy Instances,
+# we need to be fairly stingy with how much CPU/Memory we give each one.  (2)
+# This approach will significantly complicate things for Kubernets Support
+# (see below).
+#
+# - Kubernetes Support:  The PR relies on using the `docker` command line to
+# boot an Envoy container per service.  This won't work in Kubernetes where the
+# sky controller is a pod, and thus can't run sub-containers.   There are two
+# ways this could be handled.  1) Run envoy as a process instead of a docker
+# container.  2) Switch to single-envoy per controller, and boot that Envoy in
+# the pod-spec of the Kubernetes controller.
```

Comment:
It might be a little bit ugly but, can we simply add one more docker in the controller pod for each service?

Reply:
So the problem is in Kubernetes you can't dynamically add and remove containers to running pods without restarting the entire pod.  If we have a single Envoy, we can configure it to handle multiple services easily enough, but the other way around is non-trivial

Reply:
I see. makes sense. Lets try to use a unified envoy ;)